openstack/identity/v3/extensions/ec2tokens/requests.go

```diff
@@ -245,18 +265,14 @@ func (opts *AuthOptions) ToTokenV3CreateMap(map[string]interface{}) (map[string]
 	}
 
 	signedHeaders, _ := h["X-Amz-SignedHeaders"]
-	c["signature"] = EC2CredentialsBuildSignatureV4(*opts, signedHeaders, date, c["body_hash"].(string))
 
+	stringToSign := EC2CredentialsBuildStringToSignV4(*opts, signedHeaders, c["body_hash"].(string), date)
+	key := EC2CredentialsBuildSignatureKeyV4(opts.Secret, opts.Region, opts.Service, date)
+	c["signature"] = EC2CredentialsBuildSignatureV4(key, stringToSign)
 	h["X-Amz-Date"] = date.Format(EC2CredentialsTimestampFormatV4)
-	h["Authorization"] = fmt.Sprintf("%s Credential=%s/%s/%s/%s/%s, SignedHeaders=%s, Signature=%s",
-		EC2CredentialsAwsHmacV4,
-		opts.Access,
-		date.Format(EC2CredentialsDateFormatV4),
-		opts.Region,
-		opts.Service,
-		EC2CredentialsAwsRequestV4,
-		signedHeaders,
-		c["signature"])
+	h["Authorization"] = EC2CredentialsBuildAuthorizationHeaderV4(*opts, signedHeaders, c["signature"].(string), date)
+
+	c["token"] = stringToSign
```

Comment:
If my previous comment is true, should there be some kind of check to make sure this isn't already set so it's not clobbered?

Additionally, let's add a comment above this line to say that while `token` is being set, it is only used for S3 and will be removed when using EC2 validation.

Reply:
Token must be set with a signature. If signature was not provided, token will be generated as well.